src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs

```diff
@@ -501,7 +501,7 @@ public static string GetMSBuildFilePath(PackageSpec project, RestoreRequest requ
                 var projectGraph = targetGraph.Graphs.FirstOrDefault();
 
                 // Distinct union of tool packages and packages with GeneratePathProperty=true
-                var packageIdsToCreatePropertiesFor = packagesWithTools.Union(projectGraph.Item.Data.Dependencies.Where(i => i.GeneratePathProperty).Select(i => i.Name)).Distinct(StringComparer.OrdinalIgnoreCase);
+                var packageIdsToCreatePropertiesFor = new HashSet<string>(packagesWithTools.Union(projectGraph.Item.Data.Dependencies.Where(i => i.GeneratePathProperty).Select(i => i.Name)), StringComparer.OrdinalIgnoreCase);
```

Comment:
Did you consider making packagesWithTools case insensitive at line 429?  That would avoid allocating another collection.

Reply:
I have to allocate twice because you could have only wanted a property for a package for a particular TFM.

```xml
<ItemGroup>
  <PackageReference Include="Foo"
                    Version="1.0"
                    GeneratePathProperty="true"
                    Condition="'$(TargetFramework)' == 'net46'" />

  <PackageReference Include="Foo"
                    Version="1.0"
                    Condition="'$(TargetFramework)' != 'net46'" />
</ItemGroup>
```

The second `HashSet<T>` is inside the foreach loop.  Is that right @nkolev92 ?

Reply:
I was assuming that NuGet would always want to treat any collection with packageId keys as case insensitive, but I could be wrong.

Reply:
In my latest commit I'm using two different collections.  So packages with a `tools` folder get a property regardless of TFM and then a collection is allocated for each TFM for packages with `GeneratePathProperty`.  Both collections are case-insensitive.